File: backend/app/api/episodes.py
from fastapi import APIRouter, Depends, HTTPException, Request, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_session
from app.schemas.episodes import EpisodeCreate, EpisodePatch, EpisodeResponse
from app.schemas.generation import (
    GenerateAssetsResponse,
    GenerateShotsImpactResponse,
)
from app.services.episodes import (
    create_episode,
    delete_episode,
    get_episode,
    list_episodes,
    update_episode,
)
from app.services.generate_assets import enqueue_generate_assets
from app.services.generate_shots import read_generate_shots_impact
from app.tasks.queue import TaskConflictError, TaskQueue


router = APIRouter(tags=["episodes"])


@router.get(
    "/projects/{project_id}/episodes", response_model=list[EpisodeResponse]
)
async def read_episodes(
    project_id: int,
    session: AsyncSession = Depends(get_session),
) -> list[EpisodeResponse]:
    return await list_episodes(session, project_id)


@router.post(
    "/projects/{project_id}/episodes",
    response_model=EpisodeResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_episode_route(
    project_id: int,
    payload: EpisodeCreate,
    session: AsyncSession = Depends(get_session),
) -> EpisodeResponse:
    return await create_episode(session, project_id, payload)


@router.get("/episodes/{episode_id}", response_model=EpisodeResponse)
async def read_episode(
    episode_id: int,
    session: AsyncSession = Depends(get_session),
) -> EpisodeResponse:
    return await get_episode(session, episode_id)


@router.post(
    "/episodes/{episode_id}/generate-assets",
    response_model=GenerateAssetsResponse,
    status_code=status.HTTP_202_ACCEPTED,
)
async def generate_assets_route(
    episode_id: int,
    request: Request,
    session: AsyncSession = Depends(get_session),
) -> GenerateAssetsResponse:
    body = await request.body()
    if body:
        raise HTTPException(
            status_code=422, detail="Generate-assets request body must be empty"
        )

    queue: TaskQueue = request.app.state.task_queue
    try:
        result = await enqueue_generate_assets(session, queue, episode_id)
    except TaskConflictError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    await queue.publish_committed(result)
    return GenerateAssetsResponse(task_id=result.task.id)


@router.post(
    "/episodes/{episode_id}/generate-shots/impact",
    response_model=GenerateShotsImpactResponse,
)
async def generate_shots_impact_route(
    episode_id: int,
    request: Request,
    session: AsyncSession = Depends(get_session),
) -> GenerateShotsImpactResponse:
    body = await request.body()
    if body:
        raise HTTPException(
            status_code=422,
            detail="Generate-shots impact request body must be empty",
        )
    return await read_generate_shots_impact(session, episode_id)


@router.patch("/episodes/{episode_id}", response_model=EpisodeResponse)
async def update_episode_route(
    episode_id: int,
    payload: EpisodePatch,
    session: AsyncSession = Depends(get_session),
) -> EpisodeResponse:
    return await update_episode(session, episode_id, payload)


@router.delete("/episodes/{episode_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_episode_route(
    episode_id: int,
    session: AsyncSession = Depends(get_session),
) -> Response:
    await delete_episode(session, episode_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/episodes.ts
import {
  protocolError,
  requestJson,
  requestNoContent,
  requireNonNegativeSafeInteger,
  requireObjectWithKeys,
  requirePositiveSafeInteger,
  requireString,
} from "./client";

export interface Episode {
  id: number;
  project_id: number;
  seq: number;
  title: string;
  script_text: string;
  script_revision: number;
  assets_generated_script_revision: number | null;
  shots_generated_script_revision: number | null;
  created_at: string;
  updated_at: string;
}

export interface EpisodeCreate {
  seq: number;
  title: string;
  script_text?: string;
}

export interface EpisodePatch {
  seq?: number;
  title?: string;
  script_text?: string;
}

export interface GenerateAssetsResponse {
  task_id: number;
}

export interface GenerateShotsImpactResponse {
  clips_count: number;
  videos_count: number;
  confirm_token: string | null;
  expires_in: number | null;
}

export interface GenerateShotsResponse {
  task_id: number;
}

const jsonHeaders = { "Content-Type": "application/json" };

export function parseEpisodeResponse(
  value: unknown,
  status = 200,
): Episode {
  const record = requireObjectWithKeys(
    value,
    [
      "id",
      "project_id",
      "seq",
      "title",
      "script_text",
      "script_revision",
      "assets_generated_script_revision",
      "shots_generated_script_revision",
      "created_at",
      "updated_at",
    ],
    [],
    status,
    "Episode response",
  );
  requirePositiveSafeInteger(record.id, "Episode.id", status);
  requirePositiveSafeInteger(record.project_id, "Episode.project_id", status);
  requirePositiveSafeInteger(record.seq, "Episode.seq", status);
  requireString(record.title, status, "Episode.title");
  requireString(record.script_text, status, "Episode.script_text");
  requirePositiveSafeInteger(record.script_revision, "Episode.script_revision", status);
  if (record.assets_generated_script_revision !== null) {
    requireNonNegativeSafeInteger(
      record.assets_generated_script_revision,
      "Episode.assets_generated_script_revision",
      status,
    );
  }
  if (record.shots_generated_script_revision !== null) {
    requireNonNegativeSafeInteger(
      record.shots_generated_script_revision,
      "Episode.shots_generated_script_revision",
      status,
    );
  }
  requireString(record.created_at, status, "Episode.created_at");
  requireString(record.updated_at, status, "Episode.updated_at");
  return record as unknown as Episode;
}

export function parseEpisodeListResponse(
  value: unknown,
  status = 200,
): Episode[] {
  if (!Array.isArray(value)) {
    return protocolError(status, "Episode list response must be an array");
  }
  return value.map((item) => parseEpisodeResponse(item, status));
}

export function listEpisodes(projectId: number): Promise<Episode[]> {
  return requestJson<Episode[]>(`/projects/${projectId}/episodes`);
}

export function createEpisode(
  projectId: number,
  input: EpisodeCreate,
): Promise<Episode> {
  return requestJson<Episode>(`/projects/${projectId}/episodes`, {
    method: "POST",
    headers: jsonHeaders,
    body: JSON.stringify(input),
  });
}

export function getEpisode(id: number): Promise<Episode> {
  const episodeId = requirePositiveSafeInteger(id, "episodeId");
  return requestJson<Episode>(
    `/episodes/${episodeId}`,
    undefined,
    parseEpisodeResponse,
  );
}

export function updateEpisode(id: number, input: EpisodePatch): Promise<Episode> {
  return requestJson<Episode>(`/episodes/${id}`, {
    method: "PATCH",
    headers: jsonHeaders,
    body: JSON.stringify(input),
  });
}

export function generateAssets(id: number): Promise<GenerateAssetsResponse> {
  const episodeId = requirePositiveSafeInteger(id, "episodeId");
  return requestJson<GenerateAssetsResponse>(
    `/episodes/${episodeId}/generate-assets`,
    { method: "POST" },
  );
}

export function readGenerateShotsImpact(
  id: number,
): Promise<GenerateShotsImpactResponse> {
  const episodeId = requirePositiveSafeInteger(id, "episodeId");
  return requestJson<GenerateShotsImpactResponse>(
    `/episodes/${episodeId}/generate-shots/impact`,
    { method: "POST" },
  );
}

export function generateShots(
  id: number,
  confirmToken?: string,
): Promise<GenerateShotsResponse> {
  const episodeId = requirePositiveSafeInteger(id, "episodeId");
  const body = confirmToken === undefined ? {} : { confirm_token: confirmToken };
  return requestJson<GenerateShotsResponse>(
    `/episodes/${episodeId}/generate-shots`,
    {
      method: "POST",
      headers: jsonHeaders,
      body: JSON.stringify(body),
    },
  );
}

export function deleteEpisode(id: number): Promise<void> {
  const episodeId = requirePositiveSafeInteger(id, "episodeId");
  return requestNoContent(`/episodes/${episodeId}`, { method: "DELETE" });
}
